::-apple-attachment-controls-container should be gated behind ImageControlsEnabled

annevk opened the pull request:
#### 2b997ee28d1cc967b8dd04e5c453adf94ef1ba71
<pre>
::-apple-attachment-controls-container should be gated behind ImageControlsEnabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=266810">https://bugs.webkit.org/show_bug.cgi?id=266810</a>
<a href="https://rdar.apple.com/120268884">rdar://120268884</a>

Reviewed by Tim Nguyen.

As ::-apple-attachment-controls-container is the sole -apple-prefixed
pseudo-element and only used for an unstable feature we stop treating
all -apple-prefixed pseudo-elements as valid and instead special case
this specific one and preference gate it.

The impact on web developers is that they can no longer rely on
-apple-prefixed pseudo-elements to be valid in WebKit-based browsers.

Additionally remove some no-op code in the related Python script.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parsePseudoElement):
* Source/WebCore/css/SelectorPseudoElementMap.in:
* Source/WebCore/css/makeSelectorPseudoElementsMap.py:
(enumerablePseudoType):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSSelectorParserContext.h:

Canonical link: <a href="https://commits.webkit.org/272538@main">https://commits.webkit.org/272538@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd9f8e24262f55c2b8f60dd90e845174788fc2b0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32006 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10710 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33771 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34530 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29004 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32818 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13072 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7944 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28606 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32386 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/9065 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28621 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7858 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8031 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/28536 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35874 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29127 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28990 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/34137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8120 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6102 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/31995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9764 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8774 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4163 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8643 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
```diff
@@ -42,6 +42,7 @@ CSSSelectorParserContext::CSSSelectorParserContext(const CSSParserContext& conte
     , popoverAttributeEnabled(context.popoverAttributeEnabled)
     , thumbAndTrackPseudoElementsEnabled(context.thumbAndTrackPseudoElementsEnabled)
+    , imageControlsEnabled(context.imageControlsEnabled)
     , viewTransitionsEnabled(context.propertySettings.viewTransitionsEnabled)
```
Why does this code have viewTransitionsEnabled but the equivalent code in CSSParserContext.cpp does not?

nt1m replied:
`viewTransitionsEnabled` is on `propertySettings` in `CSSParserContext` which is auto-generated by CSSProperties.json

nt1m commented on Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
```diff
@@ -42,6 +42,7 @@ CSSSelectorParserContext::CSSSelectorParserContext(const CSSParserContext& conte
     , highlightAPIEnabled(context.highlightAPIEnabled)
     , popoverAttributeEnabled(context.popoverAttributeEnabled)
     , thumbAndTrackPseudoElementsEnabled(context.thumbAndTrackPseudoElementsEnabled)
+    , imageControlsEnabled(context.imageControlsEnabled)
```
This file should be sorted alphabetically.

nt1m commented on Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
```diff
@@ -56,6 +57,9 @@ CSSSelectorParserContext::CSSSelectorParserContext(const Document& document)
     , popoverAttributeEnabled(document.settings().popoverAttributeEnabled())
     , thumbAndTrackPseudoElementsEnabled(document.settings().thumbAndTrackPseudoElementsEnabled())
+#if ENABLE(SERVICE_CONTROLS)
+    , imageControlsEnabled(document.settings().imageControlsEnabled())
```
ditto here

nt1m commented on Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
```diff
@@ -72,6 +76,7 @@ void add(Hasher& hasher, const CSSSelectorParserContext& context)
         context.highlightAPIEnabled,
         context.popoverAttributeEnabled,
         context.thumbAndTrackPseudoElementsEnabled,
+        context.imageControlsEnabled,
```
ditto here

nt1m commented on Source/WebCore/css/parser/CSSSelectorParserContext.h:
```diff
@@ -44,6 +44,7 @@ struct CSSSelectorParserContext {
     bool highlightAPIEnabled { false };
     bool popoverAttributeEnabled { false };
     bool thumbAndTrackPseudoElementsEnabled { false };
+    bool imageControlsEnabled { false };
```
ditto here

nt1m commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -109,6 +109,9 @@ CSSParserContext::CSSParserContext(const Document& document, const URL& sheetBas
     , grammarAndSpellingPseudoElementsEnabled { document.settings().grammarAndSpellingPseudoElementsEnabled() }
     , customStateSetEnabled { document.settings().customStateSetEnabled() }
     , thumbAndTrackPseudoElementsEnabled { document.settings().thumbAndTrackPseudoElementsEnabled() }
+#if ENABLE(SERVICE_CONTROLS)
```
Please use `#if ENABLE(SERVICE_CONTROLS)` consistently everywhere `imageControlsEnabled` is being set/read.

nt1m reviewed (commented):
This is a good cleanup, thanks!

annevk committed:
::-apple-attachment-controls-container should be gated behind ImageControlsEnabled
https://bugs.webkit.org/show_bug.cgi?id=266810
rdar://120268884

Reviewed by Tim Nguyen.

As ::-apple-attachment-controls-container is the sole -apple-prefixed
pseudo-element and only used for an unstable feature we stop treating
all -apple-prefixed pseudo-elements as valid and instead special case
this specific one and preference gate it.

The impact on web developers is that they can no longer rely on
-apple-prefixed pseudo-elements to be valid in WebKit-based browsers.

Additionally remove some no-op code in the related Python script.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parsePseudoElement):
* Source/WebCore/css/SelectorPseudoElementMap.in:
* Source/WebCore/css/makeSelectorPseudoElementsMap.py:
(enumerablePseudoType):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSSelectorParserContext.h:

Canonical link: https://commits.webkit.org/272538@main
